### PR TITLE
Update to macro runtime & PlutoHooks.jl

### PR DESCRIPTION
**Pull request by Pangoraw**
## The companion package is [PlutoHooks.jl](https://github.com/JuliaPluto/PlutoHooks.jl)

 - [x] indirectly/directly updating the definition of a macro updates the nodes of all macros calling that macro
 - [x] directly running a macrocall updates the reactive node for that macrocall
 - [x] indirectly running a macrocall does not updates the reactive node for that macrocall
 - [x] Change computer lifecycle to use cell UUIDs instead of expression hash
 - [x] PlutoRunner hooks for cell related cleanups when the computer is removed
 - [x] PlutoRunner hooks to rerun cell implicitely
 - [x] can't use `:(function() end)` as macro outputs :cry: 
 - [x] work without function wrapping

Reactive node update also means update of the underlying function wrapper.

## Function wrapping

Cells with macro calls will be function wrapped. that means at list two things:

#### Stateful cells in Julia (similar to `this` in JavaScript)
wrapped_macrocall

<details>
<summary>Other example</summary>

wrapped_macro_acc

</details>

#### Faster execution because of a single compilation
macro_wrapped_htl

### CODE REVIEW

**Comment by rikhuijzer** [comment]
> The `@ingredients` macro is working well with #1541 so this can be used to reactively develop source file/notebooks
> 
> reactive_using .

Looks really cool. Does this work with packages too or just source files?

**Comment by Pangoraw** [comment]
> > The `@ingredients` macro is working well with #1541 so this can be used to reactively develop source file/notebooks
> 
> Looks really cool. Does this work with packages too or just source files?

Just with individual source files for now, but something that works on a package level with incremental re-definitions (using the work made on Revise.jl) is something that is definitely on the roadmap of the PlutoHooks "ecosystem". It requires to answer questions about package environment management from macros first.

File: src/runner/PlutoRunner.jl [hunk 1]
```diff
@@ -340,7 +426,14 @@ If the third argument is a `Tuple{Set{Symbol}, Set{Symbol}}` containing the refe
 
 This function is memoized: running the same expression a second time will simply call the same generated function again. This is much faster than evaluating the expression, because the function only needs to be Julia-compiled once. See https://github.com/fonsp/Pluto.jl/pull/720
 """
-function run_expression(m::Module, expr::Any, cell_id::UUID, function_wrapped_info::Union{Nothing,Tuple{Set{Symbol},Set{Symbol}}}=nothing, contains_user_defined_macrocalls::Bool=false)
+function run_expression(m::Module, expr::Any, cell_id::UUID, function_wrapped_info::Union{Nothing,Tuple{Set{Symbol},Set{Symbol}}}=nothing, forced_expr_id::Union{ObjectID,Nothing}=nothing; user_requested_run::Bool=true)
+    if user_requested_run
+        # TODO Time elapsed? Possibly relays errors in cleanup function?
+        UseEffectCleanups.trigger_cleanup(cell_id)
+
+        # TODO Could also put explicit `try_macroexpand` here, to make clear that user_requested_run => fresh macro identity
```
Comment (Pangoraw) [inline src/runner/PlutoRunner.jl:434]: no_explicit_rerun

There seems to be a problem with this, explicitely re-running does not regenerate the underlying expression. The same goes for updating the macro definition.

Comment (dralletje) [reply src/runner/PlutoRunner.jl:434]: Hmm, I don't have that D:

You didn't have that before? Maybe it's time I put the explicit re-macroexpand-ization there anyway then.

Comment (Pangoraw) [reply src/runner/PlutoRunner.jl:434]: The following condition is alway false in my example:

https://github.com/fonsp/Pluto.jl/blob/5c34b99d756e7d77d624f3155b7e9007ac1b3cc4/src/runner/PlutoRunner.jl#L453

which doesn't trigger a new expansion.

Comment (dralletje) [reply src/runner/PlutoRunner.jl:434]: Yeah, it should, as that *should* already be handled by macro_expand from the Expression Explorer.. Wonder why it isn't for you though D:

Comment (Pangoraw) [reply src/runner/PlutoRunner.jl:434]: ah yes you're right

Comment (Pangoraw) [reply src/runner/PlutoRunner.jl:434]: okay weird, i haven't been able to reproduce

**Comment by fonsp** [comment]
@Pangoraw Can you put the code for [these examples](https://github.com/fonsp/Pluto.jl/pull/1597#issuecomment-953872471) online? I want to know how to make these things!

**Comment by Pangoraw** [comment]
- Self re-rendering example: i don't have the notebook under my belt right now but the important part of the code is still visible in the gif, note that depending on the self-updating cell will re-run downstream cell even when the job is not done so doing progress with #437 will still be easier.
 - File watching: `@use_file` in PlutoHooks
 - Reactive ingredients: `@ingredients` in PlutoHooks
 - Live HTTP server response editing: currently lives in https://github.com/JuliaPluto/PlutoWebServer.jl but i have not updated it to account for the different changes yet so it might not work currently.

**Pull request merged**